### PR TITLE
[WIP] qualify in-tree scheduler plugin

### PR DESCRIPTION
**Pull request by gavinfish**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

In-tree plugin names should be fully qualified like plugin.kubescheduler.k8s.io/${pluginname} to make it easier to distinguish between in-tree and out-of-tree plugins.

**Which issue(s) this PR fixes**:

Part of  #95445

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
Users need to add prefix "plugin.kubescheduler.k8s.io/" when they use in-tree scheduler plugins.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs
None.
```

### CODE REVIEW

File: pkg/scheduler/apis/config/v1beta1/conversion.go [hunk 1]
```diff
@@ -51,7 +54,20 @@ func Convert_v1beta1_KubeSchedulerConfiguration_To_config_KubeSchedulerConfigura
 		return err
 	}
 	out.AlgorithmSource.Provider = pointer.StringPtr(v1beta1.SchedulerDefaultProviderName)
-	return convertToInternalPluginConfigArgs(out)
+	err := convertToInternalPluginConfigArgs(out)
+	if err != nil {
+		return err
+	}
+	registry := frameworkplugins.NewInTreeRegistry()
```
Comment (gavinfish) [inline pkg/scheduler/apis/config/v1beta1/conversion.go:61]: It cannot import frameworkplugins for import cycle issue, any suggestions to handle v1beta1 compatibility？@alculquicondor

Comment (alculquicondor) [reply pkg/scheduler/apis/config/v1beta1/conversion.go:61]: Yeah, we don't want to do that, even if there was no import cycle. I haven't read the PR yet, but what do you want to achieve?

The point of having the qualified names is that we don't need to check that names exist. The domain should match with the api group  https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kube-scheduler/config/v1beta1/register.go#L25

Then we can load that api group into the scheme https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/apis/config/v1beta1/conversion.go#L39
So maybe we can have a registering function.

**Comment by gavinfish** [comment]
@alculquicondor for qualifying in-tree plugin names, does it make sense to change each in-tree plugin's `Name` like in current version? Or we just need the prefix in v1beta2 api side, and remove it in conversion to use the original internal plugin name?

**Comment by alculquicondor** [comment]
If you do it here, v1beta1 configurations that use the short name will stop working. That's something we don't want to do at all.
You need to make it so that the change only affects v1beta2.

**Comment by gavinfish** [comment]
@alculquicondor PTAL, will target v1beta2 and add more ut, etc. later, but I think I am working on the right direction now?

**Comment by alculquicondor** [comment]
I don't think this approach would work for both v1beta1 and v1beta2

**Comment by gavinfish** [comment]
Some updates:
- Temporarily rebase v1beta2 api
- Add prefix "plugin.kubescheduler.k8s.io/" for each config plugins
- for v1beta1, add a temp in-tree plugin map to identify and help make compatibility for not-qualified in-tree plugin
- for v1beta2, not-qualified in-tree plugin will be taken as out-of-tree plugins
- Update related ut
- Leave some small TODO, will address later

/cc alculquicondor

**Comment by alculquicondor** [comment]
I didn't have a chance to review all the changes yet, but please you have the following tests:
- When using v1beta1 it's not necessary to have the prefix
- When using v1beta2, if there is no prefix, the plugins are not recognized as in-tree

These tests could live in scheme_test.go or options_test.go. Use your best judgement.

File: pkg/scheduler/apis/config/v1beta1/conversion.go [hunk 2]
```diff
@@ -34,6 +36,39 @@ var (
 	// Access via getPluginArgConversionScheme()
 	pluginArgConversionScheme     *runtime.Scheme
 	initPluginArgConversionScheme sync.Once
+
+	empty struct{}
```
Comment (alculquicondor) [inline pkg/scheduler/apis/config/v1beta1/conversion.go:40]: Use `sets.String` instead.

File: pkg/scheduler/apis/config/v1beta1/conversion.go [hunk 3]
```diff
@@ -73,6 +142,9 @@ func convertToInternalPluginConfigArgs(out *config.KubeSchedulerConfiguration) e
 				return fmt.Errorf("converting .Profiles[%d].PluginConfig[%d].Args into internal type: %w", i, j, err)
 			}
 			out.Profiles[i].PluginConfig[j].Args = internalArgs
+			if !strings.HasPrefix(out.Profiles[i].PluginConfig[j].Name, framework.IntreePluginPrefix) {
```
Comment (alculquicondor) [inline pkg/scheduler/apis/config/v1beta1/conversion.go:145]: Shouldn't you use `intreePluginNamesWithoutPrefixSet` here too?

Add a test for unknown plugin names not getting converted.

**Review by alculquicondor — CHANGES_REQUESTED** [review]
We also need changes in v1beta2/conversion.go

The idea is that we use the prefix to pull up the scheme for converting the plugin args.

**Comment by alculquicondor** [comment]
Hi @gavinfish, I hope you had a good time during holidays.

Can you continue working on this?

**Comment by gavinfish** [comment]
@alculquicondor some updates here.
- I have added tests for both v1beta1 and v1beta2 in scheme_test.go

I think we should talk about several items here:
- What about out-of-tree plugins, are their prefixes fixed or in the similar pattern?
- For  preFileter state key, we used to use `preFilterStateKey = "PreFilter" + Name`, once introduce prefix for in-tree plugins, it will be some thing like `PreFilterplugin.kubescheduler.k8s.io/InterPodAffinity` which is a little bit weird.

File: pkg/scheduler/apis/config/v1beta1/conversion.go (hunk 3 above)
Comment (alculquicondor) [reply pkg/scheduler/apis/config/v1beta1/conversion.go:147]: Use `intreePluginNamesWithoutPrefixSet` here too

File: pkg/scheduler/apis/config/v1beta1/conversion_test.go [hunk 4]
```diff
@@ -20,6 +20,8 @@ import (
 	"testing"
 
 	"github.com/google/go-cmp/cmp"
+
```
Comment (alculquicondor) [inline pkg/scheduler/apis/config/v1beta1/conversion_test.go:23]: remove this space

**Review by alculquicondor — COMMENTED** [review]
What about changes in v1beta2/conversion.go?

> What about out-of-tree plugins, are their prefixes fixed or in the similar pattern?

They can be arbitrary. I think we can expose a registration function where they can provide their prefix and scheme that we can use in the conversion.

> For preFileter state key

Those keys are not visible to users, so they don't matter much. But we can always change it to `pre-filter.` or remove it entirely.

**Comment by gavinfish** [comment]
> What about changes in v1beta2/conversion.go?
I have added some tests for conversion in v1beta2, do the behaviors make sense to you?

> I think we can expose a registration function where they can provide their prefix and scheme that we can use in the conversion.
Make sense to me. I will draft another PR for this. Let this PR only focus on in-tree plugins.

**Comment by alculquicondor** [comment]
> I have added some tests for conversion in v1beta2, do the behaviors make sense to you?

So basically no changes from the old v1beta1/conversion.go, correct?

> Make sense to me. I will draft another PR for this. Let this PR only focus on in-tree plugins.

That's fine.

I think this is looking good. Let's push for #95308 to be finalized first

**Pull request closed without merging**